h2/src/main/org/h2/mvstore/db/TransactionStore.java

```diff
@@ -723,6 +725,19 @@ public long setSavepoint() {
             return logId;
         }
 
+        public void markStatementStart() {
+            markStatementEnd();
+            txCounter = store.store.registerVersionUsage();
+        }
+
+        public void markStatementEnd() {
+            MVStore.TxCounter counter = txCounter;
+            txCounter = null;
+            if(counter != null) {
+                store.store.deregisterVersionUsage(counter);
+            }
+        }
+
```

Comment:
given that sessions can run on differrent threads in parallel, is this txCounter usage safe?

Reply:
While technically possible, do you really think it is a good idea to share session/transaction between threads?
How about the case when one thread commits/rolls back while another had DML operation in progress?

Reply:
my mistake, I misread the code and thought that this was shared across multiple sessions